### PR TITLE
test: making it easier to test filters for all protocols

### PR DESCRIPTION
**Pull request by alyssawilk**
Adding a test fixture which tests 
[ipv4/ipv6][http upstream/h2 upstream][http downstream/h2 downstream]

*Risk Level*: Low 
*Testing*: Exactly.
*Docs Changes*: n/a
*Release Notes*: n/a

part of #2935

Signed-off-by: Alyssa Wilk <alyssar@chromium.org>

### CODE REVIEW

File: test/integration/buffer_filter_integration_test.cc [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  testRouterRequestAndResponseWithBody(1024, 512, false);
+}
+
+TEST_P(BufferIntegrationTest, RouterRequestAndResponseWithZeroByteBodyBuffer) {
+  config_helper_.addFilter(ConfigHelper::DEFAULT_BUFFER_FILTER);
+  testRouterRequestAndResponseWithBody(0, 0, false);
+}
+
+// Verifies the typedef saves us from all ProtocolIntergrationTests from being rerun.
+TEST_P(OtherIntegrationTest, Foo) { ASSERT(0); }
```
Comment (mattklein123) [inline test/integration/buffer_filter_integration_test.cc]: nit: `ASSERT(false);`

I'm also a little confused about what this is doing. Can you expand the comment? And if this should be done in every instantiation of this test should we put in the fixture usage docs?

File: test/integration/protocol_integration.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                                public testing::TestWithParam<ProtocolTestParams> {
+public:
+  // Returns 8 combinations of
+  // [HTTP  upstream / HTTP  downstream] x [Ipv4, IPv6]
+  // [HTTP  upstream / HTTP2 downstream] x [Ipv4, IPv6]
+  // [HTTP2 upstream / HTTP2 downstream] x [IPv4, Ipv6]
+  // [HTTP upstream  / HTTP2 downstream] x [IPv4, Ipv6]
+  static std::vector<ProtocolTestParams> GetProtocolTestParams();
+
+  // Allows pretty printed test names of the form
```
Comment (mattklein123) [inline test/integration/protocol_integration.h]: I had no idea this was possible. Yes, let's do this everywhere!!!

File: test/integration/protocol_integration.h [hunk 3]
```diff
@@ -0,0 +1,53 @@
+#pragma once
+
+#include "test/integration/http_integration.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+
+struct ProtocolTestParams {
```
Comment (mattklein123) [inline test/integration/protocol_integration.h]: WDYT about `HttpProtocolTest`, http_protocol_integration.h, etc.? We have many different protocols in Envoy.

File: test/integration/protocol_integration.cc [hunk 4]
```diff
@@ -0,0 +1,31 @@
+#include "test/integration/protocol_integration.h"
+
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+std::vector<ProtocolTestParams> ProtocolIntegrationTest::GetProtocolTestParams() {
```
Comment (lizan) [inline test/integration/protocol_integration.cc]: Doesn't have to be in this PR though, can this take some parameter that limiting some combination? For example gRPC-{web,transcoder,bridge} filter integration test won't work with HTTP1 upstream.

File: test/integration/buffer_filter_integration_test.cc (hunk 1 above)
Comment (lizan) [reply test/integration/buffer_filter_integration_test.cc]: just use FAIL() from gtest :) ASSERT will be compiled out in release builds.

Comment (alyssawilk) [reply test/integration/buffer_filter_integration_test.cc]: Oops - that was supposed to have a TODO to remove.
tl;dr: will remove 

That was me making sure my typedef workaround worked.  I'm used to having one big test framework and each file has TEST_F(Foo, BarTest).  With Envoy we could do TEST_P(ProtocolngegrationTest, BarTest) and INSTANTIATE (ProtocolIntegrationTest) but I'm fairly well convinced that would play horribly badly with the giant ubertest for coverage.  We'd have 15 identical instances of INSTANTIATE-all-the-protocol-tests and everything would run 15 times. Yuck.

I wanted to make sure if we typedef both BufferIntegrationTest and OtherIntegrationTest to ProtocolTest and then INSTANTIATED the BufferIntegrationTest it would avoid the double instantiation problem.

Some day we won't have the ubertest and everything will be easier :-)

File: test/integration/protocol_integration.cc (hunk 4 above)
Comment (alyssawilk) [reply test/integration/protocol_integration.cc]: Totally.  I was going to do enums (default both) for upstream and downstream protocol, but I realized that didn't get me the combination *I* wanted which was for some tests to do which was 3 of the 4.  I'll go ahead and do that since it's useful.

Any tests which want picky things like 3 of the 4 can just set up their own vector manually.

File: test/integration/protocol_integration.h (hunk 3 above)
Comment (alyssawilk) [reply test/integration/protocol_integration.h]: Sure thing!  I'd been waffling about just colocating this with the integration test as it's not a lot of code, but forgot to ask if anyone had opinions.  HttpProtocol SGTM.
